### PR TITLE
Update aws_iid NodeAttestor plugin to use aws-sdk-go-v2

### PR DESCRIPTION
**Pull request by rturner3**
Replaces usage of AWS SDK v1 with AWS SDK v2 in an effort to eventually remove the AWS SDK v1 dependency from SPIRE
.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -78,6 +78,12 @@ require (
 	sigs.k8s.io/controller-runtime v0.11.0
 )
 
+require (
```
Comment (azdagron) [inline go.mod:81]: Hmm, is this a 3rd require section? I think we only want 2 (direct and indirect)...

File: pkg/agent/plugin/nodeattestor/aws/iid.go [hunk 2]
```diff
@@ -19,8 +21,11 @@ import (
 )
 
 const (
-	docPath = "instance-identity/document"
-	sigPath = "instance-identity/signature"
+	docPath                 = "instance-identity/document"
+	sigPath                 = "instance-identity/signature"
+	loadConfigTimeout       = 30 * time.Second
+	fetchMetadataDocTimeout = 30 * time.Second
+	fetchMetadataSigTimeout = 30 * time.Second
```
Comment (azdagron) [inline pkg/agent/plugin/nodeattestor/aws/iid.go:28]: Any reason we want individual timeouts for these operations? In general, I think we should probably leave this up to SPIRE core by threading through the context passed into the RPC handler.

File: pkg/agent/plugin/nodeattestor/aws/iid.go [hunk 3]
```diff
@@ -109,6 +118,54 @@ func fetchMetadata(endpoint string) (*caws.IIDAttestationData, error) {
...
+func getMetadataDoc(client *imds.Client) (string, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), fetchMetadataDocTimeout)
+	defer cancel()
+	res, err := client.GetDynamicData(ctx, &imds.GetDynamicDataInput{
+		Path: docPath,
+	})
+
+	if err != nil {
+		return "", err
+	}
```
Comment (azdagron) [inline pkg/agent/plugin/nodeattestor/aws/iid.go:130]: nitpick: prefer the error handling to immediately follow (keeps things grouped nicely)
```suggestion
	res, err := client.GetDynamicData(ctx, &imds.GetDynamicDataInput{
		Path: docPath,
	})
	if err != nil {
		return "", err
	}
```

File: pkg/agent/plugin/nodeattestor/aws/iid.go [hunk 4]
```diff
@@ -109,6 +118,54 @@ func fetchMetadata(endpoint string) (*caws.IIDAttestationData, error) {
...
+	return doc, nil
+}
+
+func getMetadataSig(client *imds.Client) (string, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), fetchMetadataSigTimeout)
+	defer cancel()
+	res, err := client.GetDynamicData(ctx, &imds.GetDynamicDataInput{
+		Path: sigPath,
+	})
+
```
Comment (azdagron) [inline pkg/agent/plugin/nodeattestor/aws/iid.go:146]: Same thing here.

File: pkg/agent/plugin/nodeattestor/aws/iid.go [hunk 5]
```diff
@@ -109,6 +118,54 @@ func fetchMetadata(endpoint string) (*caws.IIDAttestationData, error) {
...
+	if err != nil {
+		return "", err
+	}
+
+	sig, err := copyToString(res.Content)
+	if err != nil {
+		return "", err
+	}
+
+	return sig, nil
```
Comment (azdagron) [inline pkg/agent/plugin/nodeattestor/aws/iid.go:156]:
```suggestion
	return readStringAndClose(res.Content)
```

File: pkg/agent/plugin/nodeattestor/aws/iid.go [hunk 6]
```diff
@@ -109,6 +118,54 @@ func fetchMetadata(endpoint string) (*caws.IIDAttestationData, error) {
...
+	if err != nil {
+		return "", err
+	}
+
+	doc, err := copyToString(res.Content)
+	if err != nil {
+		return "", err
+	}
+
+	return doc, nil
```
Comment (azdagron) [inline pkg/agent/plugin/nodeattestor/aws/iid.go:137]:
```suggestion
	return readStringAndClose(res.Content)
```

File: pkg/agent/plugin/nodeattestor/aws/iid.go [hunk 7]
```diff
@@ -109,6 +118,54 @@ func fetchMetadata(endpoint string) (*caws.IIDAttestationData, error) {
...
+
+	sig, err := copyToString(res.Content)
+	if err != nil {
+		return "", err
+	}
+
+	return sig, nil
+}
+
+func copyToString(r io.ReadCloser) (res string, err error) {
```
Comment (azdagron) [inline pkg/agent/plugin/nodeattestor/aws/iid.go:159]: Can we name this something that makes it clear it takes "ownership" of the ReadCloser (i.e. closes it)?
```suggestion
func readStringAndClose(r io.ReadCloser) (res string, err error) {
```

**Review by azdagron — COMMENTED** [review]
Thanks for putting this together, @rturner3 ! I wasn't super confident that the credential loading behavior didn't change. If you get a chance the different configurations on AWS, that would go a long way :)

File: go.mod (hunk 1 above)
Comment (rturner3) [reply go.mod:81]: Not sure why a third section was added here. I will consolidate it back to two sections.

File: pkg/agent/plugin/nodeattestor/aws/iid.go (hunk 2 above)
Comment (rturner3) [reply pkg/agent/plugin/nodeattestor/aws/iid.go:28]: Not really - agreed that we should be using the stream context here

File: pkg/agent/plugin/nodeattestor/aws/iid.go [hunk 8]
```diff
@@ -109,6 +112,48 @@ func fetchMetadata(endpoint string) (*caws.IIDAttestationData, error) {
...
+	if err != nil {
+		return "", err
+	}
+
+	doc, err := readStringAndClose(res.Content)
+	if err != nil {
+		return "", err
+	}
+
+	return doc, nil
```
Comment (azdagron) [inline pkg/agent/plugin/nodeattestor/aws/iid.go:128]:
```suggestion
	return readStringAndClose(res.Content)
```

File: pkg/agent/plugin/nodeattestor/aws/iid.go [hunk 9]
```diff
@@ -109,6 +112,48 @@ func fetchMetadata(endpoint string) (*caws.IIDAttestationData, error) {
...
+	if err != nil {
+		return "", err
+	}
+
+	sig, err := readStringAndClose(res.Content)
+	if err != nil {
+		return "", err
+	}
+
+	return sig, nil
```
Comment (azdagron) [inline pkg/agent/plugin/nodeattestor/aws/iid.go:144]:
```suggestion
	return readStringAndClose(res.Content)
```

File: pkg/agent/plugin/nodeattestor/aws/iid.go [hunk 10]
```diff
@@ -109,6 +112,48 @@ func fetchMetadata(endpoint string) (*caws.IIDAttestationData, error) {
...
+
+func readStringAndClose(r io.ReadCloser) (res string, err error) {
+	defer r.Close()
+	var sb strings.Builder
+	if _, err := io.Copy(&sb, r); err == nil {
+		res = sb.String()
+	}
+
+	return
+}
```
Comment (azdagron) [inline pkg/agent/plugin/nodeattestor/aws/iid.go:155]: Sorry, I had a comment here. I think it got swallowed when I tried to submit during last weeks outage. We should avoid naked returns:
```suggestion
func readStringAndClose(r io.ReadCloser) (string, error) {
	defer r.Close()
	var sb strings.Builder
	if _, err := io.Copy(&sb, r); err == nil {
		return sb.String(), nil
	} else {
		return "", err
	}
}
```
or alternatively:
```suggestion
func readStringAndClose(r io.ReadCloser) (string, error) {
	defer r.Close()
	var sb strings.Builder
	_, err := io.Copy(&sb, r)
	return sb.String(), err
}
```

**Comment by rturner3** [comment]
> Thanks for putting this together, @rturner3 ! I wasn't super confident that the credential loading behavior didn't change. If you get a chance the different configurations on AWS, that would go a long way :)

It looks like it may take me some time to get to testing this. I also don't have a way to test the IAM role assumption case.

**Comment by rturner3** [comment]
> > Thanks for putting this together, @rturner3 ! I wasn't super confident that the credential loading behavior didn't change. If you get a chance the different configurations on AWS, that would go a long way :)
> 
> It looks like it may take me some time to get to testing this. I also don't have a way to test the IAM role assumption case.

It dawned on me that I could probably spin something up using AWS free tier resources.

**Comment by rturner3** [comment]
Thanks @MarcosDY for helping out with the testing and iteration on this PR! Marking as ready for review since this has been tested against AWS now.

**Review by azdagron — APPROVED** [review]
Huzzah!

**Pull request merged**